File: ports/csharp/src/core/kernel/BcsKernel.cs
using System;
using System.Collections.Generic;

namespace Bcs.Core.Kernel
{
    public interface IBcsModule
    {
        void Initialize();
        void Start();
        void Shutdown();
        string Name { get; }
    }

    public class BcsKernel : BcsObject
    {
        private static readonly Lazy<BcsKernel> _instance = new Lazy<BcsKernel>(() => new BcsKernel());
        public static BcsKernel Instance => _instance.Value;

        private readonly List<IBcsModule> _modules = new List<IBcsModule>();
        public BcsEventLoop EventLoop { get; }

        private bool _isStarted = false;
        private bool _isShuttingDown = false;
        private readonly object _lock = new object();

        private BcsKernel() : base(null)
        {
            EventLoop = new BcsEventLoop(new BcsEventDispatcher(), this);
        }

        public void RegisterModule(IBcsModule module)
        {
            lock (_lock)
            {
                if (!_isStarted)
                {
                    _modules.Add(module);
                }
            }
        }

        public void Startup()
        {
            lock (_lock)
            {
                if (_isStarted) return;
                _isStarted = true;
            }

            // Copy to avoid locks during callbacks
            List<IBcsModule> modulesCopy;
            lock (_lock) { modulesCopy = new List<IBcsModule>(_modules); }

            foreach (var mod in modulesCopy)
            {
                Console.WriteLine($"BcsKernel: Initializing module {mod.Name}");
                mod.Initialize();
            }

            foreach (var mod in modulesCopy)
            {
                Console.WriteLine($"BcsKernel: Starting module {mod.Name}");
                mod.Start();
            }
        }

        public int Exec()
        {
            try
            {
                Startup();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"BcsKernel Startup Error: {ex.Message}");
                return -1;
            }

            Console.WriteLine("BcsKernel: Entering main event loop...");
            int code = EventLoop.Exec();

            Shutdown();
            return code;
        }

        public void Shutdown()
        {
            lock (_lock)
            {
                if (_isShuttingDown) return;
                _isShuttingDown = true;
            }

            Console.WriteLine("BcsKernel: Shutting down...");

            List<IBcsModule> modulesCopy;
            lock (_lock) { modulesCopy = new List<IBcsModule>(_modules); }

            for (int i = modulesCopy.Count - 1; i >= 0; i--)
            {
                var mod = modulesCopy[i];
                Console.WriteLine($"BcsKernel: Shutting down module {mod.Name}");
                try { mod.Shutdown(); } catch { /* Ignore cleanup errors */ }
            }

            Dispose();
        }
    }
}

File: ports/java/src/core/kernel/BcsKernel.java
package core.kernel;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicBoolean;

public class BcsKernel extends BcsObject {

    public interface BcsModule {
        void initialize() throws Exception;
        void start() throws Exception;
        void shutdown();
        String getName();
    }

    private static volatile BcsKernel instance;
    private final List<BcsModule> modules = new CopyOnWriteArrayList<>();
    private final BcsEventLoop eventLoop;

    public BcsEventLoop getEventLoop() {
        return eventLoop;
    }


    private final AtomicBoolean isStarted = new AtomicBoolean(false);
    private final AtomicBoolean isShuttingDown = new AtomicBoolean(false);

    private BcsKernel() {
        super(null);
        this.eventLoop = new BcsEventLoop(new BcsEventDispatcher(), this);
    }

    public static BcsKernel getInstance() {
        if (instance == null) {
            synchronized (BcsKernel.class) {
                if (instance == null) {
                    instance = new BcsKernel();
                }
            }
        }
        return instance;
    }

    public void registerModule(BcsModule module) {
        if (!isStarted.get()) {
            modules.add(module);
        }
    }

    public void startup() throws Exception {
        if (!isStarted.compareAndSet(false, true)) {
            return;
        }

        // Phase 1: Initialize
        for (BcsModule mod : modules) {
            System.out.println("BcsKernel: Initializing module " + mod.getName());
            mod.initialize();
        }

        // Phase 2: Start
        for (BcsModule mod : modules) {
            System.out.println("BcsKernel: Starting module " + mod.getName());
            mod.start();
        }
    }

    public int exec() {
        try {
            startup();
        } catch (Exception ex) {
            System.out.println("BcsKernel Startup Error: " + ex.getMessage());
            return -1;
        }

        System.out.println("BcsKernel: Entering main event loop...");
        int code = eventLoop.exec();

        shutdown();
        return code;
    }

    public void shutdown() {
        if (!isShuttingDown.compareAndSet(false, true)) {
            return;
        }

        System.out.println("BcsKernel: Shutting down...");

        List<BcsModule> modulesCopy = new ArrayList<>(modules);
        for (int i = modulesCopy.size() - 1; i >= 0; i--) {
            BcsModule mod = modulesCopy.get(i);
            System.out.println("BcsKernel: Shutting down module " + mod.getName());
            try {
                mod.shutdown();
            } catch (Exception e) {
                // Ignore cleanup errors
            }
        }

        close();
    }
}
